Sources/AlgoliaSearch-Client/FilterBuilder/Filters/FilterFacet.swift

```diff
@@ -0,0 +1,131 @@
+//
+//  FilterFacet.swift
+//  AlgoliaSearch OSX
+//
+//  Created by Vladislav Fitc on 27/12/2018.
+//  Copyright © 2018 Algolia. All rights reserved.
+//
+
+import Foundation
+
+/// Defines facet filter
+///[Filter by string](https:www.algolia.com/doc/guides/managing-results/refine-results/filtering/how-to/filter-by-string/)
+///[Filter by boolean](https:www.algolia.com/doc/guides/managing-results/refine-results/filtering/how-to/filter-by-boolean/)
+public struct FilterFacet: Filter, Hashable {
+    
+    public let attribute: Attribute
+    public let value: ValueType
+    public var isInverted: Bool
+    
+    public let expression: String
+    
+    public init(attribute: Attribute, value: ValueType, isInverted: Bool = false) {
+        self.attribute = attribute
+        self.isInverted = isInverted
+        self.value = value
+        self.expression = """
+        "\(attribute)":"\(value)"
+        """
+    }
+    
+    public init(_ facetTuple: FacetTuple) {
+        self.init(attribute: facetTuple.0, value: facetTuple.1)
+    }
+    
+    public init(attribute: Attribute, stringValue: String, isInverted: Bool = false) {
+        self.init(attribute: attribute, value: .string(stringValue), isInverted: isInverted)
+    }
+    
+    public init(attribute: Attribute, floatValue: Float, isInverted: Bool = false) {
+        self.init(attribute: attribute, value: .float(floatValue), isInverted: isInverted)
+    }
+    
+    public init(attribute: Attribute, boolValue: Bool, isInverted: Bool = false) {
+        self.init(attribute: attribute, value: .bool(boolValue), isInverted: isInverted)
+    }
+    
+    public func replacingAttribute(by attribute: Attribute) -> FilterFacet {
+        return FilterFacet(attribute: attribute, value: value, isInverted: isInverted)
+    }
+    
+}
+
+extension FilterFacet: RawRepresentable {
+
+    public typealias RawValue = (Attribute, ValueType)
+    
+    public init?(rawValue: (Attribute, FilterFacet.ValueType)) {
+        self.init(attribute: rawValue.0, value: rawValue.1)
+    }
+    
+    public var rawValue: (Attribute, FilterFacet.ValueType) {
+        return (attribute, value)
+    }
+    
+}
+
+extension FilterFacet {
+    
+    // TODO: The getter of valueType is a bit heavy, but we want to keep the context of the type used in the init (bool, float, string).
```

Comment:
Is this still valid, and can it be solved now?

Reply:
It cannot. We don't want to separate FilterFacet with different nested values to different types, as in this case we cannot put facet filters with different nested types in the same disjunctive groups. Now we can just declare it as OrGroup<FilterFacet> and put in it any FilterFacet instance. If we separate them, OrGroup<StringFilterFacet> and OrGroup<FloatFilterFacet> (or if you prefer OrGroup<FilterFacet<String>> and OrGroup<FilterFacet<Float>>) are not compatible anymore. 
The best we can do, is to introduce an optional getter for each possible type:
var boolValue: Bool?
var floatValue: Float?
var stringValue: Float?
A bit NSNumber-like way, but the best solution so far, as we want one type which rules them all.
What do you think?

Reply:
Okay got it, let's keep as it 👍